fw_upgrade/README.md

```diff
@@ -0,0 +1,123 @@
+# Firmware Upgrade
+
+Upgrading the router's firmware usually works via NetCloud, for newer firmware releases (`> 2018`) this is the only way. For older releases upgrading via `scp` or the web interface still works.
+
+Older firmware images are still available on https://cradlepoint.com. They are signed with a RSA private key and encrypted with AES256-CBC. However, the signature verification can be easily bypassed and the AES key is hard-coded in the router's Python middleware.
+
+## Implementation
+
+The firmware upgrade is mainly implemented in the `/service_manager/services/utils/upgrade_write.pyc` file. This is what's going on:
+
+1. `upgrade_write()` --> decrypts the upgrade file, writes a block, gets the `uimage` and `uboot` signatures from the firmware file *or* calls `upgrade_write_init()`
+2. `upgrade_write_init()` --> header and CRC checks, writes a new SHA512 hash to flash memory
+3. `_write()` --> writes data to memory flash
+4. `upgrade_validate()` --> validates the file and checks the signature by calling `upgrade_validate_signature()`
+
+If the version string in the firmware header matches a certain value, the signature verification can be bypassed:
```

Comment:
maybe better write "signature is ignored" than "signature verification can be bypassed"

Reply:
Fixed.